### PR TITLE
Integrated circuits for modular computers

### PR DESCRIPTION
**Pull request by Ghommie**
## About The Pull Request
This PR integrates circuits for modular computers and a good bits of their programs.
The peculiarity here is that modular computers have no fixed amount of unremovable components (except the base one with just a couple ports for now), instead, they're added and removed along with programs. With a few exceptions (such as the messenger and signaler), for these program circuits to work, their associated program has to be either open or in the background.

For a reason or another, not all programs have a circuit associated to them, still, however the programs with a circuit are still a handful. They are:
- Nanotrasen Pay System
- Notepad
- SiliConnect
- WireCarp
- MODsuit Control
- Spectre Meter
- Direct Messenger*
- LifeConnect
- Custodial Locator
- Fission360
- Camera
- Status Display
- SignalCommander

*By the by, sending messages has a cooldown, so it shouldn't be as spammy. If it turns out to not be enough, I can make it so messages from circuit will be ignored by other messenger circuits.

The PR is no longer WIP.

## Why It's Good For The Game
I believe modular computers could make for some interesting setups with circuits, since they're fairly flexible and stocked with features unlike many other appliances, therefore also a speck more abusable, though limits, cooldowns, logging and sanitization have been implemented to keep it in check.

## Changelog

:cl:
add: Modular Computers now support integrated circuits. What can be done with them depends on the programs installed and whether they're running (open or background).
add: Modular Consoles (the machinery) now have a small backup cell they draw power from if the power goes out.
/:cl:

### CODE REVIEW

**Comment by MTandi** [comment]
I jumped on my chair at first because I thought that you made the circuits [programmable with a laptop...](https://hackmd.io/@ZSLHD7nTTlWNegwCVVdJBw/S1MdVTF43)

Still more things to integrate with is better.

I see that you have added a lot of circuit interfaces, can I request adding a usb port to APC to be able to turn night light, machinery power, etc via usb?

**Comment by dieamond13** [comment]
I should be able to mine crypto using circuits and modular computers

**Comment by moocowswag** [comment]
If this pr is merged ill come back to the game to explore the "balance" of this.

**Comment by Ghommie** [comment]
This should be ready for review now.

**Comment by Sakamoto4ka** [comment]
I think about stuff like rust modular puties in indigo

**Comment by moocowswag** [comment]
ERROR: File(s) with no trailing newline detected, please add one.

:(

File: code/modules/modular_computers/file_system/program_circuit.dm [hunk 1]
```diff
@@ -0,0 +1,63 @@
+
+
```
Comment (Time-Green) [inline code/modules/modular_computers/file_system/program_circuit.dm:2]: thats a lotta empty space

File: code/modules/modular_computers/file_system/programs/borg_monitor.dm [hunk 2]
```diff
@@ -164,3 +183,31 @@
 
 /datum/computer_file/program/borg_monitor/syndicate/checkID()
 	return "\[CLASSIFIED\]" //no ID is needed for the syndicate version's message function, and the borg will see "[CLASSIFIED]" as the message sender.
+
+
```
Comment (Time-Green) [inline code/modules/modular_computers/file_system/programs/borg_monitor.dm:187]: empty space

**Review by Time-Green — CHANGES_REQUESTED** [review]
Yep, I understand circuit code

**Comment by Time-Green** [comment]
PR says WIP, but seems ready otherwise?

**Comment by Ghommie** [comment]
Mhm.

**Comment by moocowswag** [comment]
what are the "targets" for the messenger program? they are entity variables but it doesnt work when you input a scanned person or pda and use it as the target.

**Comment by moocowswag** [comment]
Okay so the only thing of mechanical value is that lifeline lets you track people's coordinates remotely for use with launchpad.

**Comment by Ghommie** [comment]
> what are the "targets" for the messenger program? they are entity variables but it doesnt work when you input a scanned person or pda and use it as the target.

The "targets" port is a list of entities, it , so you need to wrap the entity in a list first. It could be also not working as intended, I'll double-check it.

**Pull request merged**